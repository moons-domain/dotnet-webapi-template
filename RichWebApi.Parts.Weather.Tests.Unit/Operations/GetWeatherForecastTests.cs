using MediatR;
using Microsoft.Extensions.DependencyInjection;
using RichWebApi.Dependencies.Database.Tests;
using RichWebApi.Entities;
using RichWebApi.Persistence;
using RichWebApi.Tests;
using RichWebApi.Tests.DependencyInjection;
using RichWebApi.Tests.Resources;
using Xunit.Abstractions;

namespace RichWebApi.Operations;

public class GetWeatherForecastTests : UnitTest
{
	private readonly IResourceScope _resourceScope;
	private readonly IServiceProvider _serviceProvider;

	public GetWeatherForecastTests(ITestOutputHelper testOutputHelper, ResourceRepositoryFixture resourceRepository, DependencyContainerFixture dependencyContainerFixture) : base(testOutputHelper)
	{
		_resourceScope = resourceRepository.BeginTestScope(this);
		var parts = new AppPartsCollection()
			.AddWeather();
		_serviceProvider = dependencyContainerFixture
			.WithXunitLogging(testOutputHelper)
			.WithTestScopeInMemoryDatabase(this, parts)
			.WithSystemClock()
			.ConfigureServices(s => s.ApplyParts(parts));
	}

	[Fact]
	public async Task LoadsFirstPage()
	{
		var entities = _resourceScope.GetJsonInputResource<WeatherForecast[]>("entities");
		var db = _serviceProvider.GetRequiredService<IRichWebApiDatabase>();
		foreach (var wf in entities)
		{
			await db.Context.AddAsync(wf);
		}

		await db.PersistAsync();
		
		
		var input = _resourceScope.GetJsonInputResource<GetWeatherForecast>();
		var mediator = _serviceProvider.GetRequiredService<IMediator>();
		
		var result = await mediator.Send(input);
		_resourceScope.CompareWithJsonExpectation(TestOutputHelper, result);
	}
}