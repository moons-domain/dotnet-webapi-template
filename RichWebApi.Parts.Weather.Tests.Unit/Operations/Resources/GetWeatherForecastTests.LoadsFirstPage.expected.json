{
  "Items": [
    {
      "Date": "2023-01-01T00:00:00",
      "TemperatureC": 14,
      "TemperatureF": 57,
      "Summary": "Summary"
    },
    {
      "Date": "2023-01-02T00:00:00",
      "TemperatureC": 7,
      "TemperatureF": 44,
      "Summary": "Cold"
    },
    {
      "Date": "2023-01-03T00:00:00",
      "TemperatureC": -3,
      "TemperatureF": 27,
      "Summary": "Very cold"
    }
  ],
  "Page": 0,
  "Size": 10,
  "Total": 3
}